### PR TITLE
Add transform between cover and media & text

### PR DESCRIPTION
**Pull request by walbo**
## Description
Add transform between cover and media & text

## Testing Instructions

1. Insert a Media & Text block and fill with content.
2. Try to transform it to a Cover block.
3. Transform it back to Media & Text

## Screenshots 
media-cover-tranform

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 
- [ ] I've updated related schemas if appropriate.

### CODE REVIEW

**Review by apeatling — CHANGES_REQUESTED** [review]
This is a great transform!

One issue I found when testing is the text color. When setting the text color on the media and text block, it is not preserved when doing a transform back to media and text from the cover.

File: packages/block-library/src/media-text/transforms.js
Comment (aaronrobertshaw) [inline packages/block-library/src/media-text/transforms.js]: My understanding is there's a general preference for reducing our use of Lodash, with the view to one day drop it entirely. I think we can avoid the use of `set` here.

**Review by aaronrobertshaw — COMMENTED** [review]
These transforms make a lot of sense. Nice work! 👍 

Like @apeatling, I had mixed results when a text color was selected. It appears the main issue is one block opts into color support for text and the other relies on its inner blocks to provide that customization.

Transforming a Media & Text block to a cover block loses the color selection immediately for me. If instead of setting the text color on the Media & Text block, you set it on the inner paragraph, then the color is maintained given the transform doesn't modify the inner blocks.

**Comment by walbo** [comment]
> One issue I found when testing is the text color. When setting the text color on the media and text block, it is not preserved when doing a transform back to media and text from the cover.

Yeah, I'm not sure how to solve this since the cover block uses the `isDark` attribute to set a classname that defines the color. Any suggestions?

File: packages/block-library/src/media-text/transforms.js
Comment (walbo) [reply packages/block-library/src/media-text/transforms.js]: Yep. Will refactor the code to drop the use of lodash. Didn't know there was an effort to use less lodash 👍

**Comment by glendaviesnz** [comment]
> > One issue I found when testing is the text color. When setting the text color on the media and text block, it is not preserved when doing a transform back to media and text from the cover.
> 
> Yeah, I'm not sure how to solve this since the cover block uses the `isDark` attribute to set a classname that defines the color. Any suggestions?

I think it would be difficult to fix this without refactoring the Cover block to have a text color option as well as overlay color, and it would make sense to have it opt into the color block supports for this. 

If we were to do this I think that would be a separate PR to this one. It seems like this is a small enough limitation to be able to merge this PR knowing this, and knowing that it will be followed up later once Cover has text color support?

**Comment by aaronrobertshaw** [comment]
> I think it would be difficult to fix this without refactoring the Cover block to have a text color option as well as overlay color, and it would make sense to have it opt into the color block supports for this.

@glendaviesnz there could be another interim option to this, which I've been exploring this morning. We could have the transform propagate the text color selections onto the inner blocks. If the inner blocks don't opt into color support it has no effect however if the inner block does opt in, the user's color selections are maintained. 

It needs more testing and further thought but it could reduce the impact of the text color issue and allow these transforms to land sooner. 

I'll put together a separate PR extending this one so it can be discussed further in isolation.

**Comment by aaronrobertshaw** [comment]
As per my last comment, I've created https://github.com/WordPress/gutenberg/pull/40666 to demo the possibility of applying text color selections to inner blocks.

It appears to work Media & Text > Cover > Media & Text. When converting back the inner blocks retain the color selections rather than them being moved back to the Media & Text block, I think this is ok for now. 

Interested to hear what everyone thinks, in particular, about honoring the user made color selections over the `isDark` class etc. I'm inclined to think keeping things looking as the user expects and selected is a good thing.

To move the addition of these transforms forward, I see a few options:

1. Push ahead with this PR as is, accepting the confusing UX when text color selections have been made.
2. Apply alternate approach in [#40666](https://github.com/WordPress/gutenberg/pull/40666) and proceed
3. Hold off on these transforms until we explore the Cover block opting into color block support (even if only for text color)

**Comment by apeatling** [comment]
I don't think we should hold off because I think this is a beneficial transform even with the loss of text color. I noticed that your PR raised some issues, if we can resolve those then this combined with that change seems like a good approach.

**Comment by aaronrobertshaw** [comment]
>  I noticed that your PR raised some issues

I'm not sure we can make the UX perfect when transforming Media & Text > Cover > Media & Text. 

I didn't find a means of identifying when we copied the text color onto the inner block versus when the user had set them on the inner block in the initial Media & Text.

That said, it appears we're all leaning towards getting these transforms merged. Mostly likely with https://github.com/WordPress/gutenberg/pull/40666 applied first.

If the Cover block eventually opts into text color block support, the transform can be simplified then.

@walbo are you happy with the above plan? Any thoughts or concerns?

**Comment by walbo** [comment]
@aaronrobertshaw Sound good to me. Added a comment on #40666

**Review by aaronrobertshaw — APPROVED** [review]
I've merged in the changes from https://github.com/WordPress/gutenberg/pull/40666 and given this another test. 

Looks to be working well for me though it might be good to get another approval from @apeatling or @glendaviesnz given the most recent changes made to this PR were mine.

**Review by apeatling — APPROVED** [review]
LGTM 👍

**Pull request merged**
